### DISCUSSION TITLE
How do conditional fragments affect the cache?

### DISCUSSION THREAD

**Discussion by nandorojo** [urql-graphql/urql#2651] [Q&A] [answered]
Hi there! I have a question about conditional fragments. I didn't see anything mentioned in the docs about them, so I thought I would ask here.

Say I have a query like this:

```graphql
query Artist($slug: String!) {
  artist(slug: $slug) {
    ... on Artist {
     id
     name
   }
   ... on ArtistDoesNotExist {
     exists
     deletedAt
   }
  }
}
```

I want to query an artist, but also allow for the condition that this artist does not exist. Currently, we are throwing an error when this happens.

However, I'd like to show a unique UI when the `ArtistDoesNotExist`, rather than a plain `error`. So I figured I would use GraphQL's conditional fragments feature.

I wanted to ask: **how does this work with a normalized cache?** If I have a successful `Artist` query cached for `artist(slug:drake)`, will that continue to pull from the cache in the future with those variables? 

Since I typically query `artist` with _either_ `slug` or `id`, [I use a resolver for the `artist` node for `slug` and `id`](https://github.com/FormidableLabs/urql/discussions/2519#discussioncomment-3015242). 

But I'm not sure what `cache.resolve('Query', 'artist', { slug, id })` will return if I use conditional fragments.

Is there anything special I should know since the return type could be either `Artist` or `ArtistDoesNotExist`?

Please let me know if I should clarify any of this. Thank you as always!

**Comment by JoviDeCroock** [comment, marked as answer]
One thing to note is that this approach of differentiating between errors and an actual result isn't spec-compliant, I see the appeal of using unions for strict-typing but throwing this out there. This means that if this errors out we will be caching this error 😅 I see that this might be desired for a notFound but thought I would make you aware of it just to be sure.

So yes, you will just keep pulling from the cache for _both_ the success as well as the error response, the result will be saved as any other successful query. In one case `query.Artist(id: x) --> NotFound` it will be embedded data as it isn't keyable and in the other `query.Artist(id: y) --> Artist:y`.

This does mean that you need a `keys: { NotFound: () => null }`

Generally I like to give the advice to use the normal graphql-errors to avoid running into weird client-cache scenario's where artist(id: x) --> null accompanied with some `graphQLErrors`.

**Reply by nandorojo** [reply]
> One thing to note is that this approach of differentiating between errors and an actual result isn't spec-compliant.

Good to know. For example, what if a user [is banned on Twitter](https://twitter.com/eIonmusk)? I looked in their network tab, and I see a GraphQL response that closely mirrors what I am describing. 

<img width="1590" alt="Screen Shot 2022-08-30 at 2 42 03 PM" src="https://user-images.githubusercontent.com/13172299/187517415-ca96cf5b-ca87-47dd-b590-71ce579a7711.png">

> PS this isn't Elon, it has a capital "i" instead of an L haha

Note that it has `__typename: 'UserUnavailable'` with a reason and messages. This perfectly mirrors what I am trying to do. After all, a missing/deleted artist is not an unknown error that is caused by some server malfunction.

Plus, the GraphQL conditional fragments seem like such a good opportunity to provide much better experiences for predictable cases where we know certain entities won't resolve, since they were deleted or blocked.

**Reply by nandorojo** [reply]
> This does mean that you need a keys: { NotFound: () => null }

That makes sense, I was planning on doing that. Good to know.

**Reply by nandorojo** [reply]
> Generally I like to give the advice to use the normal graphql-errors to avoid running into weird client-cache scenario's where artist(id: x) --> null accompanied with some graphQLErrors.

This is good to know. To clarify, do you mean that if `data` for `artist(id: 123)` returns `null` due to some `error`, it could cause problems for the rest of the cache?

Thanks for all the insight! Really helpful!

**Reply by nandorojo** [reply]
Based on all this, it sounds like I might be best off returning an error. I'll play with it to see if there is anything to be learned, and I'll share here if I come across any unusual behavior.

**Comment by nandorojo** [comment]
In case anyone is curious, I found this useful article from a Twitter employee (Sasha Solomon) about 200 errors: https://sachee.medium.com/200-ok-error-handling-in-graphql-7ec869aec9bc

She recommends using fragments the way I do. It could be useful to have a guide on best-practices/gotchas for this kind of approach with URQL's normalized cache. I might just try it out and see how it goes...

**Comment by kitten** [comment]
Hiya, First and foremost this will just work. If your core question is how it works; it'll become embedded data, i.e. if your union has an entity that may have no key, then the link (i.e. the field) will point to a keyless piece of data.

I have an article that fully explains how every part of Graphcache works here: https://kitten.sh/graphql-normalized-caching

On a side note, if your related question is how you'd model keyed entities that can be represented by unions _with keys_, for instance for authentication, then you could create an edge type

**Reply by nandorojo** [reply]
> I have an article that fully explains how every part of Graphcache works here: https://kitten.sh/graphql-normalized-caching

I reference this often. It's a great piece of writing and very insightful. 

> On a side note, if your related question is how you'd model keyed entities that can be represented by unions with keys, for instance for authentication, then you could create an edge type

To make sure I understand what you mean here, are you suggesting making an edge type like this?

```graphql
artist(id: 123) {
  node { name, id, slug }
  deleted { timestamp }
}
```

**Reply by kitten** [reply]
Basically, I was referencing types like these:

```graphql
type User {
  id: ID!
  name: String
}

type UserError {
  // ...
}

type UserEdge {
  id: ID
  node: User!
}

union UserResult = UserEdge | UserError
```

**Reply by nandorojo** [reply]
Thanks for clarifying. I was planning on doing this:

```graphql
type User {
  id: ID!
  name: String
}

type UserError {
  // ...
}

union UserResult = User | UserError
```

I assume it's the same idea, since `UserResult` is a union in both cases, where only one of them has a key.

**Reply by kitten** [reply]
Yes, it's the same from a GraphQL response's perspective. Just to clarify again, you'd add an edge if you want to let a normalised cache reuse this field result across multiple responses

**Reply by nandorojo** [reply]
Hi @kitten, I'm revisiting this point and implementing your suggestion. To clarify, if I want to use inline fragments, I should organize the backend to return a response like this:

```js
// User
response = {
  id: '123',
  __typename: 'UserEdge',
  node: {    
    __typename: 'User',
    id: '123',
    name: 'Fernando'
  }
}


// UserError
response = {
  id: '123',
  __typename: 'UserEdge',
  node: {    
    __typename: 'UserError',
    message: 'User does not exist.'
  }
}
```

Am I right to use the same `id` field for both `UserEdge` and for `User`? My guess would be that they don't necessarily have to be the same, but rather that they have to be consistent across queries for the same user.

The example I gave is inconsistent with what you sent, which is `union User = UserEdge | UserError`. So is the right answer this second method?

```js
// User
response = {
  id: '123',
  __typename: 'UserEdge',
  node: {    
    __typename: 'User',
    id: '123',
    name: 'Fernando'
  }
}


// UserError
response = { 
    __typename: 'UserError',
    message: 'User does not exist.' 
}
```

**Reply by kitten** [reply]
@nandorojo: The edge doesn't necessarily need an ID or a key field. An edge is mostly useful because it just says "A is related to B". This means that it's commonly unique to a key that's more akin to A's key + B's key.

In many cases you can get away with no key hence, because it's likely that the edge is only ever appearing in "one place" meaning, there are no cases where the edge needs to update or does update. For example, in a "result edge" the `User` may update but the edge doesn't have to know about this.

It's more common to give the edge a key if it appears in multiple places, e.g. for edges that represent parent-child relations, where the edge may appear both on a parent entity and child entity.

I'd suspect however that in your case you don't need to give the edge any keyable field. If you for instance have `Query.artist(...)` then the field itself is a sufficiently unique key, if it's the only access point to this same edge and you only ever need it to be resolved from a prior query.
